### DISCUSSION TITLE
Add method hooks to core

### DISCUSSION THREAD

**Discussion by jamauro** [meteor/meteor#12363] [Feature Requests]
I think it'd be nice to have a built-in way to perform functions `before` and `after` write methods -- `insert`, `update`, `upsert`, `remove` -- without having to extend each collection.

The `before` hooks would be useful to validate data for example. Currently I'm intercepting the write methods and injecting my own validation functions. It works but feels less than ideal.

I looked at the `collection-hooks` package but it seems to have a lot of reported issues both in the repo and the forums. I'm assuming that it suffers from the same issue of lack of an official way to have these hooks so it resorts to monkey patching.

If these hooks were exposed, I'm not sure how you'd ensure that if multiple packages used them, they wouldn't conflict. But maybe that's a non-issue.

**Comment by henriquealbert** [comment]
Have you tried https://github.com/Grubba27/meteor-rpc?

**Reply by jamauro** [reply]
I looked at it and think it’s interesting but I’m not sure I want to be locked into using zod. It also wasn’t clear from the docs if I’d still be able to get the stub value like with Validated Methods. I think the create module feature is particularly interesting in that package. I like the idea of having all db functions namespaced with the collection name.

**Reply by radekmie** [reply]
There's also [`zodern:relay`](https://github.com/zodern/meteor-relay) (it supports stubs). You don't have to use them if you don't want to rely on Zod, but they are a sensible inspiration to build on.

**Comment by radekmie** [comment]
There was already a discussion about it in https://github.com/meteor/meteor/pull/11039 and I don't think it's something that should be a part of the Core. Both (Meteor) method- and collection-level hooks are implementable in the user space.
